### ISSUE TITLE
OSX: TcpListener.Start() doesn't throw if the port is already in use

### ISSUE THREAD

**Issue by IlyaBiryukov** [dotnet/corefx#32707]
OS: Mac OS High Sierra 10.13.6
DotNet Core 2.1.402

**Repro Steps:**

1. Create and run a following NodeJS program to listen on port 8080:
```javascript
var http = require('http');

http.createServer(function (req, res) {
    res.write('Hello world!');
    res.end();
}).listen(8080);
```

2. Connect to `localhost:8080` in browser and ensure that this server works.

3. Now create and run a dotnet core 2.1.402 console app program that would attempt to listen on the same port on localhost:
```csharp
var listener = new TcpListener(IPAddress.Loopback, 8080);
listener.Start();
var socket = listener.AcceptSocket();
```

*Expected:*
`SocketException` is thrown because port 8080 is already in use.

*Actual:*
Nothing happens. The tcp listener is created but it will never accept a connection. All connections in browser go to the previously started NodeJS app.
However, if you don't run the NodeJs app first, the dotnet server will happily accept the connection.

This is a regression from some 2.0 version of dotnet core on OSX, and inconsistent with Windows
where listener.Start() throws as expected if the port is in use.

**Comment by stephentoub** [comment]
cc: @tmds

**Comment by tmds** [comment]
> Expected:
SocketException is thrown because port 8080 is already in use.

This has the expected behavior on Linux.

> The tcp listener is created but it will never accept a connection. All connections in browser go to the previously started NodeJS app.

I guess the .NET Core app may be bound to 127.0.0.1 and nodejs on ::1. Can you access the .NET Core app on 127.0.0.1?

> This is a regression from some 2.0 version of dotnet core on OSX

With what .NET Core 2.0 version does this work as expected?

**Comment by IlyaBiryukov** [comment]
On Linux, Ubuntu 16.04, `listener.Start()` does throw `SocketException` in this scenario. 

On Mac dotnet app binds to localhost, and nodejs app - to *. Connecting to 127.0.0.1 in browser connects to both dotnet app and to the nodejs app on Mac and also on Windows.

On Mac and Windows if NodeJs app starts the server with `listen(8080, 'localhost')` then dotnet's `new TcpListener(IPAdress.Loopback, 8080).Start();` throws as expected. The issue only repros when NodeJS starts the server on any interfaces, with just `listen(8080)`. In this case, only `new TcpListener(IPAddress.Any).Start()` or ...IPAddress.IPv6Any... throw but not ...IPAdress.Loopback...

I thought it was a regression from .NET Core 2.0; now I can reproduce the issue on Mac with dotnet 2.0 as well.

Other issue I see on Mac is that `IPGlobalProperties.GetIPGlobalProperties().GetActiveTcpListeners()` does not list NodeJS app listener in the above scenario. It's not the case for Windows and Linux. I've filed #32727 for that.

**Comment by wfurt** [comment]
man 7 socket 
> For  AF_INET sockets this means that a socket may bind, except when
> there is an active listening socket bound to the address.

I agree with @tmds that running service on same port but different addresses should be ok. 
Note that on OSX I can get even multiple sockets on same port AND same address.
(running multiple instances of nc -l 6666)

```
macik2:~ furt$ netstat -an | grep 6666
tcp4       0      0  *.6666                 *.*                    LISTEN
tcp4       0      0  *.6666                 *.*                    LISTEN
```

However the observed .net behavior seems right. Listener should fail if it binds to same port & same address. 

#32727 can be fixed but relying on TCP list has race condition and therefore it will never be reliable. 
Only bind() is atomic as it is handled by kernel.  

It seems like we should close this. (and fix #32727)

**Comment by tmds** [comment]
This can be closed.
